File: tests/java_helics/TestValueFederate.java
import java.lang.reflect.Method;
import java.util.*;

import com.java.helics.SWIGTYPE_p_void;
import com.java.helics.SWIGTYPE_p_double;
//import com.java.helics.SWIGTYPE_p_int64_t;
import com.java.helics.helics;
import com.java.helics.helicsConstants;
import com.java.helics.federate_state;

public class TestValueFederate {
	public static SWIGTYPE_p_void AddBroker(String core_type) {
		String initstring = "1 --name=mainbroker";
		String version = helics.helicsGetVersion();
		// Create broker
		SWIGTYPE_p_void broker = helics.helicsCreateBroker(core_type, "", initstring);
		int isconnected = helics.helicsBrokerIsConnected(broker);
		assert isconnected == 1;
		return broker;
	}

	public static SWIGTYPE_p_void AddFederate(SWIGTYPE_p_void broker, String core_type,
			int count, double deltat, String name_prefix) {

	    // Create Federate Info object that describes the federate properties #
	    SWIGTYPE_p_void fedinfo = helics.helicsCreateFederateInfo();

	    // Set core type from string
	    helics.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq");

	    // Federate init string
	    String fedinitstring = "--broker=mainbroker --federates="+ Integer.toString(count);
	    helics.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring);

	    // Set one second message interval
	    helics.helicsFederateInfoSetTimeProperty(fedinfo, 137, deltat);

	    helics.helicsFederateInfoSetIntegerProperty(fedinfo, 271, 1);

	    SWIGTYPE_p_void vFed = helics.helicsCreateValueFederate(name_prefix + "TestA Federate", fedinfo);

	    return vFed;
	}

	public static void test_value_federate_initialize(SWIGTYPE_p_void vFed) {
		federate_state status = helics.helicsFederateGetState(vFed);
		assert status.swigValue() == 0;
		helics.helicsFederateEnterExecutingMode(vFed);
		status = helics.helicsFederateGetState(vFed);
		assert status.swigValue() == 3;
	}

	public static void test_value_federate_publisher_registration(SWIGTYPE_p_void vFed) {
		SWIGTYPE_p_void pubid1 = helics.helicsFederateRegisterTypePublication(vFed, "pub1",
								 "string", "");
		SWIGTYPE_p_void pubid2 = helics.helicsFederateRegisterGlobalTypePublication(vFed, "pub2",
				 "int", "");
		SWIGTYPE_p_void pubid3 = helics.helicsFederateRegisterTypePublication(vFed, "pub3",
				 "double", "V");
		helics.helicsFederateEnterExecutingMode(vFed);
	}

	public static void test_value_federate_publication_registration(SWIGTYPE_p_void vFed) {
		SWIGTYPE_p_void pubid1 = helics.helicsFederateRegisterPublication(vFed, "pub1",
				 helicsConstants.HELICS_DATA_TYPE_STRING, "");
		SWIGTYPE_p_void pubid2 = helics.helicsFederateRegisterGlobalPublication(vFed, "pub2",
				helicsConstants.HELICS_DATA_TYPE_INT, "");
		SWIGTYPE_p_void pubid3 = helics.helicsFederateRegisterPublication(vFed, "pub3",
				helicsConstants.HELICS_DATA_TYPE_DOUBLE, "V");
		helics.helicsFederateEnterExecutingMode(vFed);

	    int maxLen = 100;
	    String pubKey = helics.helicsPublicationGetKey(pubid1);
	    assert pubKey == "TestA Federate/pub1";
	    String pubKey2 = helics.helicsPublicationGetKey(pubid2);
		assert pubKey2 == "pub2";
	    String pubKey3 = helics.helicsPublicationGetKey(pubid3);
		assert pubKey3 == "TestA Federate/pub3";
	    String pub_type = helics.helicsPublicationGetType(pubid3);
		assert pub_type == "double";
	    String pub_units = helics.helicsPublicationGetUnits(pubid3);
	    assert pub_units == "V";
	}

	// public static void test_value_federate_subscription_registration(SWIGTYPE_p_void vFed) {
	// 	SWIGTYPE_p_void subid1 = helics.helicsFederateRegisterOptionalSubscription(vFed, "sub1", "double", "V");
	// 	SWIGTYPE_p_void subid2 = helics.helicsFederateRegisterOptionalTypeSubscription(vFed, "sub2",
	// 			helicsConstants.HELICS_DATA_TYPE_INT, "");
	// 	SWIGTYPE_p_void subid3 = helics.helicsFederateRegisterOptionalSubscription(vFed, "sub3", "double", "V");
	// 	helics.helicsFederateEnterExecutingMode(vFed);
	// }

	public static void test_value_federate_subscription_and_publication_registration(SWIGTYPE_p_void vFed) {
		SWIGTYPE_p_void pubid1 = helics.helicsFederateRegisterTypePublication(vFed, "pub1",
				"string", "");
		SWIGTYPE_p_void pubid2 = helics.helicsFederateRegisterGlobalTypePublication(vFed, "pub2",
				"int", "");

		SWIGTYPE_p_void pubid3 = helics.helicsFederateRegisterPublication(vFed, "pub3", 
				helicsConstants.HELICS_DATA_TYPE_INT, "V");

		// SWIGTYPE_p_void subid1 = helics.helicsFederateRegisterOptionalSubscription(vFed, "sub1", "double", "V");
		// SWIGTYPE_p_void subid2 = helics.helicsFederateRegisterOptionalTypeSubscription(vFed, "sub2",
		// 		helicsConstants.HELICS_DATA_TYPE_INT, "");

		// SWIGTYPE_p_void subid3 = helics.helicsFederateRegisterOptionalSubscription(vFed, "sub3",
		// 		"double", "V");
	}

	public static void test_value_federate_single_transfer(SWIGTYPE_p_void vFed) {
		SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalTypePublication (vFed, "pub1",
	    		"string", "");
		SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");

	    helics.helicsFederateEnterExecutingMode(vFed);

	    helics.helicsPublicationPublishString(pubid, "string1");

	    double requestTime = 0.0;
	    double timeout = helics.helicsFederateRequestTime(vFed, requestTime);
	    byte[] s = new byte[100];
		int[] len = new int[1];
	    helics.helicsInputGetString(subid, s, len);
		String sub_string = new String(s);
	    assert sub_string == "string1";
	}

	public static void test_value_federate_runFederateTestDouble(SWIGTYPE_p_void vFed) {
	    double defaultValue = 1.0;
	    double testValue = 2.0;
	    SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalTypePublication (vFed, "pub1",
	    		"double", "");
	    SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");
	    helics.helicsInputSetDefaultDouble(subid, defaultValue);

	    helics.helicsFederateEnterExecutingMode (vFed);

	    // publish string1 at time=0.0;
	    helics.helicsPublicationPublishDouble(pubid, testValue);

	    double value = helics.helicsInputGetDouble(subid);
	    assert value == defaultValue;

	    double requesttime = 0.0;
	    double timeout = helics.helicsFederateRequestTime (vFed, requesttime);
	    assert requesttime == 0.1;

	    value = helics.helicsInputGetDouble(subid);
	    assert value == testValue;

	    // publish string1 at time=0.0;
	    helics.helicsPublicationPublishDouble(pubid, testValue + 1);
	    
	    timeout = helics.helicsFederateRequestTime (vFed, requesttime);
	    assert requesttime == 0.02;

	    value = helics.helicsInputGetDouble(subid);
	    assert value == testValue + 1;
	}

	public static void test_value_federate_runFederateTestComplex(SWIGTYPE_p_void vFed) {
	    double rDefaultValue = 1.0;
	    double iDefaultValue = 1.0;
	    double rTestValue = 2.0;
	    double iTestValue = 2.0;
	    SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalTypePublication (vFed, "pub1",
	    		"complex", "");
	    SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");
	    helics.helicsInputSetDefaultComplex(subid, rDefaultValue, iDefaultValue);

	    helics.helicsFederateEnterExecutingMode (vFed);

	    // publish string1 at time=0.0;
	    helics.helicsPublicationPublishComplex(pubid, rTestValue, iTestValue);
	    double[] value1 = new double[1];
	    double[] value2 = new double[1];
	    helics.helicsInputGetComplex(subid, value1, value2);
	    assert value2[0] == iDefaultValue;

	    double requesttime = 0.0;
	    double timeout = helics.helicsFederateRequestTime (vFed, requesttime);
	    assert requesttime == 0.01;

	    helics.helicsInputGetComplex(subid, value1, value2);
	    assert value1[0] == rTestValue;
	    assert value2[1] == iTestValue;
	}


	public static void test_value_federate_runFederateTestInteger(SWIGTYPE_p_void vFed) {
	    double defaultValue = 1;
	    double testValue = 2;
	    SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalTypePublication (vFed, "pub1", "int", "");
	    SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");
	    //SWIGTYPE_p_int64_t pDf = null;
	    //helics.helicsSubscriptionSetDefaultInteger(subid, pDf);

	    helics.helicsFederateEnterExecutingMode (vFed);
	    //SWIGTYPE_p_int64_t pValue = null;
	    //helics.helicsPublicationPublishInteger(pubid, pValue);

	    long value = helics.helicsInputGetInteger(subid);
	    assert value == defaultValue;
	    double requesttime = 0.0;
	    double timeout = helics.helicsFederateRequestTime(vFed, requesttime);
	    assert requesttime == 0.01;

	    value = helics.helicsInputGetInteger(subid);
	    assert value == testValue;

	    //helics.helicsPublicationPublishInteger(pubid, pValue);
	    timeout = helics.helicsFederateRequestTime (vFed, requesttime);
	    assert requesttime == testValue;

	    value = helics.helicsInputGetInteger(subid);
	    assert value == testValue + 1;
	}

	public static void test_value_federate_runFederateTestString(SWIGTYPE_p_void vFed) {
	    String defaultValue = "String1";
	    String testValue = "String2";
	    SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalTypePublication (vFed, "pub1",
	    		"string", "");
	    SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");
	    helics.helicsInputSetDefaultString(subid, defaultValue);

	    helics.helicsFederateEnterExecutingMode(vFed);

	    // TODO: Fix error with the following function
	    helics.helicsPublicationPublishString(pubid, testValue);
	    byte[] value = new byte[100];
		int[] len = new int[1];
	    helics.helicsInputGetString(subid, value, len);
		String val = new String(value);
	    assert val == defaultValue;
	    double requesttime = 0.0;
	    double timeout = helics.helicsFederateRequestTime (vFed, requesttime);
	    assert requesttime == 0.01;

	    helics.helicsInputGetString(subid, value, len);
		String val2 = new String(value);
	    assert val2 == testValue;
	}

	public static void test_value_federate_runFederateTestVectorD(SWIGTYPE_p_void vFed) {
	    double[] defaultValue = {0, 1, 2};
	    double[] testValue = {3, 4, 5};
	    int len = 0;
	    SWIGTYPE_p_void pubid = helics.helicsFederateRegisterGlobalInput (vFed, "pub1", helicsConstants.HELICS_DATA_TYPE_VECTOR, "");
	    SWIGTYPE_p_void subid = helics.helicsFederateRegisterSubscription (vFed, "pub1", "");
	    helics.helicsInputSetDefaultVector(subid, defaultValue, 3);

	    helics.helicsFederateEnterExecutingMode(vFed);

	    // TODO: Fix error with the following function
	    double[] data = new double[100];
	    helics.helicsPublicationPublishVector(pubid, data, 100);

	    int[] actualLen = new int[1];
		SWIGTYPE_p_double pData = null;
	    helics.helicsInputGetVector(subid, pData, 3,actualLen);
	    //assert value == [0, 1, 2];
	    double requesttime = 0.0;
	    double timeout = helics.helicsFederateRequestTime(vFed, requesttime);
	    assert requesttime == 0.01;

	    double[] value = new double[1];
	    helics.helicsInputGetVector(subid, pData, 3,actualLen);
	    //assert value == [3, 4, 5]
	}

	public static void main(String[] args) {
		SWIGTYPE_p_void broker = AddBroker("zmq");
		String initstring = "--broker=";
		String id = helics.helicsBrokerGetIdentifier(broker);
		initstring = initstring + id;
		initstring = initstring + " --broker_address";
		String address = helics.helicsBrokerGetAddress(broker);
		helics.helicsBrokerDisconnect(broker);
		helics.helicsCloseLibrary();
	}

}

File: tests/python/test_value_federate.py
import os
import time
import pytest as pt
import helics as h


@pt.fixture
def vFed():

    initstring = "-f 1 --name=mainbroker"
    fedinitstring = "--broker=mainbroker --federates=1"
    deltat = 0.01
    # TODO: should add an assert here about helicsGetVersion
    h.helicsGetVersion()

    # Create broker #
    broker = h.helicsCreateBroker("zmq", "", initstring)

    isconnected = h.helicsBrokerIsConnected(broker)

    if isconnected == 1:
        pass

    # Create Federate Info object that describes the federate properties #
    fedinfo = h.helicsCreateFederateInfo()

    # Set Federate name #
    h.helicsFederateInfoSetCoreName(fedinfo, "TestA Core")

    # Set core type from string #
    h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

    # Federate init string #
    h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

    # Set the message interval (timedelta) for federate. Note th#
    # HELICS minimum message time interval is 1 ns and by default
    # it uses a time delta of 1 second. What is provided to the
    # setTimedelta routine is a multiplier for the default timedelta.

    # Set one second message interval #
    h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)

    h.helicsFederateInfoSetIntegerProperty(fedinfo, h.helics_property_int_log_level, 1)

    vFed = h.helicsCreateValueFederate("TestA Federate", fedinfo)

    yield vFed

    h.helicsFederateFinalize(vFed)

    state = h.helicsFederateGetState(vFed)
    assert state == 3

    while h.helicsBrokerIsConnected(broker):
        time.sleep(1)

    h.helicsFederateInfoFree(fedinfo)
    h.helicsFederateFree(vFed)
    h.helicsCloseLibrary()


def test_error_value_federate_initialize():
    initstring = "-f 1 --name=mainbroker"
    with pt.raises(h._helics.HelicsException):
        h.helicsCreateBroker("mq", "", initstring)


def test_value_federate_initialize(vFed):
    state = h.helicsFederateGetState(vFed)
    assert state == 0

    h.helicsFederateEnterExecutingMode(vFed)

    state = h.helicsFederateGetState(vFed)
    assert state == 2


def test_value_federate_publication_registration(vFed):
    pubid1 = h.helicsFederateRegisterTypePublication(vFed, "pub1", "string", "")
    pubid2 = h.helicsFederateRegisterGlobalTypePublication(vFed, "pub2", "int", "")
    pubid3 = h.helicsFederateRegisterTypePublication(vFed, "pub3", "double", "V")
    h.helicsFederateEnterExecutingMode(vFed)

    publication_key = h.helicsPublicationGetKey(pubid1)
    assert publication_key == "TestA Federate/pub1"
    publication_key = h.helicsPublicationGetKey(pubid2)
    assert publication_key == "pub2"
    publication_key = h.helicsPublicationGetKey(pubid3)
    assert publication_key == "TestA Federate/pub3"
    publication_type = h.helicsPublicationGetType(pubid3)
    assert publication_type == "double"
    publication_units = h.helicsPublicationGetUnits(pubid3)
    assert publication_units == "V"


def test_value_federate_runFederateTestNamedPoint(vFed):
    defaultValue = (
        "start of a longer string in place of the shorter one and now this should be very long"
    )
    defVal = 5.3
    # testValue1 = "inside of the functional relationship of helics"
    testValue1 = "short string"
    testVal1 = 45.7823
    testValue2 = "I am a string"
    testVal2 = 0.0

    pubid = h.helicsFederateRegisterGlobalPublication(
        vFed, "pub1", h.helics_data_type_named_point, ""
    )
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")

    h.helicsInputSetDefaultNamedPoint(subid, defaultValue, defVal)

    h.helicsFederateEnterExecutingMode(vFed)

    # publish string1 at time=0.0;
    h.helicsPublicationPublishNamedPoint(pubid, testValue1, testVal1)

    # double val;
    val = h.helicsInputGetNamedPoint(subid)
    # assert value == defaultValue
    assert val == [defaultValue, defVal]

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)

    assert grantedtime == 0.01

    # get the value
    val2 = h.helicsInputGetNamedPoint(subid)
    # make sure the string is what we expect
    # assert value2 == testValue1
    assert val2 == [testValue1, testVal1]

    # publish a second string
    h.helicsPublicationPublishNamedPoint(pubid, testValue2, testVal2)

    # make sure the value is still what we expect
    val3 = h.helicsInputGetNamedPoint(subid)
    # make sure the string is what we expect
    # assert value3 == testValue1
    assert val3 == [testValue1, testVal1]

    # advance time
    grantedtime = h.helicsFederateRequestTime(vFed, 2.0)
    assert grantedtime == 0.02

    # make sure the value was updated
    val4 = h.helicsInputGetNamedPoint(subid)
    # make sure the string is what we expect
    # assert value4 == testValue2
    assert val4 == [testValue2, testVal2]


def test_value_federate_runFederateTestBool(vFed):
    defaultValue = True
    testValue1 = True
    testValue2 = False

    # register the publications
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_boolean, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")

    h.helicsInputSetDefaultBoolean(subid, h.helics_true if defaultValue else h.helics_false)

    h.helicsFederateEnterExecutingMode(vFed)

    # publish string1 at time=0.0;
    h.helicsPublicationPublishBoolean(pubid, h.helics_true if testValue1 else h.helics_false)
    val = h.helicsInputGetBoolean(subid)

    assert val == h.helics_true if defaultValue else h.helics_false

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    # get the value
    val = h.helicsInputGetBoolean(subid)

    # make sure the string is what we expect
    assert val == h.helics_true if testValue1 else h.helics_false

    # publish a second string
    h.helicsPublicationPublishBoolean(pubid, h.helics_true if testValue2 else h.helics_false)

    # make sure the value is still what we expect
    val = h.helicsInputGetBoolean(subid)
    assert val == h.helics_true if testValue1 else h.helics_false
    # advance time
    grantedtime = h.helicsFederateRequestTime(vFed, 2.0)
    # make sure the value was updated
    assert grantedtime == 0.02

    val = h.helicsInputGetBoolean(subid)
    assert val == h.helics_false if testValue2 else h.helics_true


def test_value_federate_publisher_registration(vFed):
    pubid1 = h.helicsFederateRegisterPublication(vFed, "pub1", h.helics_data_type_string, "")
    pubid2 = h.helicsFederateRegisterGlobalPublication(vFed, "pub2", h.helics_data_type_int, "")
    pubid3 = h.helicsFederateRegisterPublication(vFed, "pub3", h.helics_data_type_double, "V")
    h.helicsFederateEnterExecutingMode(vFed)

    publication_key = h.helicsPublicationGetKey(pubid1)
    assert publication_key == "TestA Federate/pub1"
    publication_type = h.helicsPublicationGetType(pubid1)
    assert publication_type == "string"
    publication_key = h.helicsPublicationGetKey(pubid2)
    assert publication_key == "pub2"
    publication_key = h.helicsPublicationGetKey(pubid3)
    assert publication_key == "TestA Federate/pub3"
    publication_type = h.helicsPublicationGetType(pubid3)
    assert publication_type == "double"
    publication_units = h.helicsPublicationGetUnits(pubid3)
    assert publication_units == "V"
    publication_type = h.helicsPublicationGetType(pubid2)
    assert publication_type == "int64"


def test_value_federate_subscription_and_publication_registration(vFed):
    pubid3 = h.helicsFederateRegisterTypePublication(vFed, "pub3", "double", "V")

    subid1 = h.helicsFederateRegisterSubscription(vFed, "sub1", "")
    subid2 = h.helicsFederateRegisterSubscription(vFed, "sub2", "")
    subid3 = h.helicsFederateRegisterSubscription(vFed, "sub3", "V")

    h.helicsFederateEnterExecutingMode(vFed)

    publication_type = h.helicsPublicationGetType(pubid3)
    assert publication_type == "double"

    sub_key = h.helicsSubscriptionGetKey(subid1)
    assert sub_key == "sub1"
    sub_type = h.helicsInputGetType(subid1)
    assert sub_type == ""
    sub_key = h.helicsSubscriptionGetKey(subid2)
    assert sub_key == "sub2"
    sub_key = h.helicsSubscriptionGetKey(subid3)
    assert sub_key == "sub3"
    sub_type = h.helicsInputGetType(subid3)
    assert sub_type == ""
    sub_units = h.helicsInputGetUnits(subid3)
    assert sub_units == "V"
    sub_type = h.helicsInputGetType(subid2)
    assert sub_type == ""


def test_value_federate_single_transfer(vFed):

    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_string, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")

    h.helicsFederateEnterExecutingMode(vFed)

    h.helicsPublicationPublishString(pubid, "string1")

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    s = h.helicsInputGetString(subid)
    assert s == "string1"


def test_value_federate_runFederateTestDouble(vFed):
    defaultValue = 1.0
    testValue = 2.0
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_double, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultDouble(subid, defaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    # publish string1 at time=0.0;
    h.helicsPublicationPublishDouble(pubid, testValue)

    value = h.helicsInputGetDouble(subid)
    assert value == defaultValue

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    value = h.helicsInputGetDouble(subid)
    assert value == testValue

    # publish string1 at time=0.0;
    h.helicsPublicationPublishDouble(pubid, testValue + 1)

    grantedtime = h.helicsFederateRequestTime(vFed, 2.0)
    assert grantedtime == 0.02

    value = h.helicsInputGetDouble(subid)
    assert value == testValue + 1


def test_value_federate_runFederateTestComplex(vFed):
    rDefaultValue = 1.0
    iDefaultValue = 1.0
    rTestValue = 2.0
    iTestValue = 2.0
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_complex, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultComplex(subid, rDefaultValue, iDefaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    # publish string1 at time=0.0;
    h.helicsPublicationPublishComplex(pubid, rTestValue, iTestValue)

    value1, value2 = h.helicsInputGetComplex(subid)
    assert value1 == rDefaultValue
    assert value2 == iDefaultValue

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    value1, value2 = h.helicsInputGetComplex(subid)
    assert value1 == rTestValue
    assert value2 == iTestValue


def test_value_federate_runFederateTestInteger(vFed):
    defaultValue = 1
    testValue = 2
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_int, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultInteger(subid, defaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    h.helicsPublicationPublishInteger(pubid, testValue)

    value = h.helicsInputGetInteger(subid)
    assert value == defaultValue

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    value = h.helicsInputGetInteger(subid)
    assert value == testValue

    h.helicsPublicationPublishInteger(pubid, testValue + 1)

    grantedtime = h.helicsFederateRequestTime(vFed, 2.0)
    assert grantedtime == 0.02

    value = h.helicsInputGetInteger(subid)
    assert value == testValue + 1


def test_value_federate_runFederateTestString(vFed):
    defaultValue = "String1"
    testValue = "String2"
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_string, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultString(subid, defaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    h.helicsPublicationPublishString(pubid, testValue)

    value = h.helicsInputGetString(subid)
    assert value == defaultValue

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    value = h.helicsInputGetString(subid)
    assert value == testValue


def test_value_federate_runFederateTestVectorD(vFed):
    defaultValue = [0, 1, 2]
    testValue = [3, 4, 5]
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_vector, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultVector(subid, defaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    h.helicsPublicationPublishVector(pubid, testValue)

    value = h.helicsInputGetVector(subid)
    assert value == [0, 1, 2]

    grantedtime = h.helicsFederateRequestTime(vFed, 1.0)
    assert grantedtime == 0.01

    value = h.helicsInputGetVector(subid)
    assert value == [3, 4, 5]


@pt.fixture
def helicsBroker():
    initstring = "-f 1 --name=mainbroker"
    # TODO: should add an assert here about helicsGetVersion
    h.helicsGetVersion()

    # Create broker #
    broker = h.helicsCreateBroker("zmq", "", initstring)

    isconnected = h.helicsBrokerIsConnected(broker)

    if isconnected == 1:
        pass

    yield broker

    h.helicsBrokerFree(broker)
    h.helicsCloseLibrary()


@pt.mark.skipif(
    os.getenv("HELICS_PYTHON_ADDITIONAL_TESTS", "") == "", reason="Disabled by default."
)
def test_value_federate_runFederateTimeoutTest(helicsBroker):

    fedinitstring = "--broker=mainbroker --federates=1"
    deltat = 0.01

    # Create Federate Info object that describes the federate properties #
    fedinfo = h.helicsCreateFederateInfo()

    # Set Federate name #
    h.helicsFederateInfoSetCoreName(fedinfo, "TestA Core")

    # Set core type from string #
    h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

    # Federate init string #
    h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

    # Set the message interval (timedelta) for federate. Note th#
    # HELICS minimum message time interval is 1 ns and by default
    # it uses a time delta of 1 second. What is provided to the
    # setTimedelta routine is a multiplier for the default timedelta.

    # Set one second message interval #
    h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)

    h.helicsFederateInfoSetIntegerProperty(fedinfo, h.helics_property_int_log_level, 1)

    vFed = h.helicsCreateValueFederate("TestA Federate", fedinfo)

    defaultValue = "String1"
    testValue = "String2"
    pubid = h.helicsFederateRegisterGlobalPublication(vFed, "pub1", h.helics_data_type_string, "")
    subid = h.helicsFederateRegisterSubscription(vFed, "pub1", "")
    h.helicsInputSetDefaultString(subid, defaultValue)

    h.helicsFederateEnterExecutingMode(vFed)

    counter = 60
    while counter > 0:
        counter -= 1
        time.sleep(1)

    # Broker should be connected at this point
    assert h.helicsBrokerIsConnected(helicsBroker) == 1, "Broker should still be connected"

    h.helicsFederateFinalize(vFed)

    state = h.helicsFederateGetState(vFed)
    assert state == 3

    while h.helicsBrokerIsConnected(helicsBroker):
        time.sleep(1)

    h.helicsFederateInfoFree(fedinfo)
    h.helicsFederateFree(vFed)
